Solution pool

ZedongPeng opened the pull request:
## Summary/Motivation:
Add the support of **solution pool** in MindtPy.

## Finished
- support of solution pool in CPLEX

## TODO

- [x] support of GUROBI
- [x] add tests for solution pool
- [ ] check solution pool works well for OA/GOA multitree (benchmark)
- [ ] restructure the MindtPy_iteration_loop function


### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -127,6 +127,11 @@ def _get_MindtPy_config():
     ))
+    CONFIG.declare('num_solution_iteration', ConfigValue(
+        default=5,
+        description='The number of MIP solutions(from solution pool) used to generate fixed NLP subproblem per iteration.',
```
Correct as "The number of MIP solutions (from the solution pool) used to generate the fixed NLP subproblem in each iteration."

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -488,3 +493,6 @@ def check_config(config):
                 'The threads parameter is corrected to 1 since incumbent callback conflicts with multi-threads mode.')
+    if config.solution_pool:
+        if config.mip_solver not in {'cplex_persistent', 'gurobi_persistent'}:
+            config.mip_solver = 'cplex_persistent'
```
Why do you prefer cplex as the default?

ZedongPeng replied:
The reason is that I first implemented CPLEX solution pool. Actually, I don't have any preference. We can wait for the benchmark results and then decide.

bernalde commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -23,6 +23,7 @@
 from pyomo.opt import SolverFactory
 from pyomo.common.dependencies import attempt_import
 from pyomo.contrib.gdpopt.util import copy_var_list_values
+from pyomo.solvers.plugins.solvers.gurobi_direct import gurobipy
```
Shouldn't this be an attempt import?

ZedongPeng replied:
gurobi is **attempt imported** in `pyomo.solvers.plugins.solvers.gurobi_direct` . That's why I don't `attempt_import gurobipy` here. Should we rely on the `attempt_import` code in gurobi_direct or directly `attempt_import gurobi` here?

bernalde commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -97,25 +98,80 @@ def MindtPy_iteration_loop(solve_data, config):
+                            solve_data.working_model)
+                        if solve_data.curr_int_sol in set(solve_data.integer_list):
+                            config.logger.info(
+                                'The same combination has been explored and will be skipped here.')
```
THis is just a question, wouldn't it be useful to tell the user which combinations have been previously explored? Maybe not under logger.info but under logger.debug

ZedongPeng replied:
I remember we tried to print the detailed combination if cycling happened (eg. [1, 0, 0, 1, 1, ...]). Actually, we cannot get any information from the numbers. So we choose to tell the user this combination has been explored in which previous iteration. I'll do so here.

ZedongPeng replied:
But since the solution pool is activated, we cannot easily index the store `integer_list` and get the iteration in which the same combination has been explored.

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_global_lp_nlp.py:
```diff
@@ -1,27 +1,15 @@
 # -*- coding: utf-8 -*-
 """Tests for the MindtPy solver."""
```
Add some description of what this tests are for (global lp nlp)

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_solution_pool.py:
```diff
@@ -0,0 +1,61 @@
+"""Tests for the MindtPy solver."""
```
Add test description (solution pool)

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_solution_pool.py:
```diff
@@ -0,0 +1,61 @@
+if all(SolverFactory(s).available() for s in required_solvers):
+    subsolvers_available = True
+else:
+    subsolvers_available = False
```
A test is feailing here with the following error:
``pyomo.common.errors.ApplicationError: No Python bindings available for <class 'pyomo.solvers.plugins.solvers.cplex_persistent.CPLEXPersistent'> solver plugin``
I would also suggest splitting the tests, sometimes CPLEX might be available but Gurobi not (or the opposite) so we want to run all the tests that we can.

bernalde commented on pyomo/contrib/mindtpy/tests/unit_test.py:
```diff
@@ -9,22 +9,13 @@
 #  ___________________________________________________________________________
 
 """Tests for the MindtPy solver."""
```
Same comment as before, add test description (unit tests)

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -22,6 +22,8 @@
 from pyomo.core.expr.calculus.derivatives import differentiate
 from pyomo.common.dependencies import attempt_import
 from pyomo.contrib.fbbt.fbbt import fbbt
+from math import fabs
```
Do we still need fabs at some point?

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -606,3 +608,47 @@ def setup_solve_data(model, config):
+            if ignore_integrality \
+                    and v_to.is_integer():  # not v_to.is_continuous()
+                v_to.value = var_val
+            elif v_to.is_integer() and (fabs(var_val - rounded_val) <= config.integer_tolerance):  # not v_to.is_continuous()
```
Change this to abs() since it is faster, type stable and works for this case better.

bernalde reviewed (commented):
Great PR with a lot of cleanup and newer implementation.
What I do not quite understand is whether you want to include the other elements in the todo list as part of this PR ( check solution pool works well for OA/GOA multitree (benchmark)
and restructure the MindtPy_iteration_loop function).
I think that both are important since they will make us discover possible bugs but are not strictly necessary.
Thank you for keeping working on this!

jsiirola committed:
GAMS writer efficiency update: add special handling of LinearExpression

jsiirola committed:
GAMS/Baron: more special handling for MonomialTerms

emma58 committed:
Fixing a couple bugs, deprecating transforming BooleanVars that aren't used in expressions, but still doing it for backwards compatibility.

emma58 committed:
Making LogicalConstraint targets possible

emma58 committed:
Cleaning up, moving some of the functions onto the transformation

emma58 committed:
Removing an unused variable

emma58 committed:
Adding a test of indexed logical constraint targets

jsiirola committed:
Do not include space when negating a monomial term

jsiirola committed:
Efficiency improvements for LinearExpression

jsiirola committed:
Moving print statements into test error message

jsiirola committed:
Disable implicit repn cache for LP,NL writers

jsiirola committed:
Remove reference to _repn

jsiirola committed:
Remove reference to _repn

jsiirola committed:
NFC: updating docstring to reflect updated LinearExpression.args format

emma58 committed:
Merge branch 'main' into l-to-l-targets

jsiirola committed:
Resolve LinearExpression backward copatibility, add tests

emma58 committed:
Addings tests for transforming blocks

emma58 committed:
No need to do a separate loop to find the BooleanVars--doing that during the LogicalConstraint transformations

emma58 committed:
Adding a test for a LogicalConstraintData target

jsiirola committed:
Adding tests for obj/con repn caching in LP / NL writers

jsiirola committed:
Update Pyros tests to skip additional tests that require baron

jsiirola committed:
Skip additional GDP test when glpk is not available

jsiirola committed:
Bugfix for slicing (map 'inf' do None)

jsiirola committed:
Fixing typo

emma58 committed:
Merge pull request #2135 from emma58/logical-to-linear

Logical_to_linear on hierarchical models

emma58 committed:
Merge branch 'main' into l-to-l-targets

jsiirola committed:
Merge branch 'main' into expression-replacement

jsiirola committed:
Merge pull request #2146 from jsiirola/test-without-solvers

Resolve test failures when no solvers are available

jsiirola committed:
Improce code comments, test coverage

Robbybp committed:
remove unused branch parsing output from slice_component_along_sets

Robbybp committed:
nfc: remove whitespace

Robbybp committed:
fix typo

Robbybp committed:
check that user-provided descend indices are valid

Robbybp committed:
test to cover StopIteration when trying to descend with an empty slice

jsiirola committed:
Merge pull request #2144 from jsiirola/disable-implicit-repn-cache

Disable implicit repn cache for LP,NL writers

jsiirola committed:
Add ordered argument to IndexedComponent keys/values/items

Robbybp committed:
change argument name to normalize and add inline comments explaining why it is necessary

Robbybp committed:
todo comment

ZedongPeng committed:
resolve comments

ZedongPeng committed:
available exception_flag=False

ZedongPeng commented:
> Great PR with a lot of cleanup and newer implementation. What I do not quite understand is whether you want to include the other elements in the todo list as part of this PR ( check solution pool works well for OA/GOA multitree (benchmark) and restructure the MindtPy_iteration_loop function). I think that both are important since they will make us discover possible bugs but are not strictly necessary. Thank you for keeping working on this!

Since the PR usually takes some time, I think we can submit the PR to pyomo main branch if you think this PR is good. 
For the other elements in the todo list, we can do a small benchmark for the solution pool in the meantime of the PR. Just to check solution pool works well. I tried to do restructure the `MindtPy_iteration_loop` function but failed.

michaelbynum committed:
tests for differentiation with nested expressions

jsiirola committed:
Fix keys() for Param coponents

jsiirola committed:
Fix initialization of ordered Set with unordered data (set())

jsiirola committed:
Improve iterator efficiency

jsiirola committed:
Update documentation, improve iterator efficiency

jsiirola committed:
Add test for keys(ordered=True)

Robbybp committed:
run unittest.main instead of specific tests

Robbybp committed:
fix typo

blnicho committed:
Merge pull request #2149 from jsiirola/ordered-keys

Add `ordered=True` API for iterating over IndexedComponents

blnicho committed:
Merge pull request #2141 from Robbybp/flatten-update

Utility function to slice individual components in flatten module

jsiirola committed:
Define DataPortalInitializer

jsiirola committed:
Fix typo (originally identified in #2118)

jsiirola committed:
Allow passing Initializers to Initializer

jsiirola committed:
NFC: PEP8

michaelbynum committed:
better handling of repeated sub-expressions

jsiirola committed:
Adding a missing test guard

Robbybp committed:
transfer variable values instead of writing new nlp

Robbybp committed:
remove unused ComponentMap import

jsiirola committed:
Initializer should treat Series as sequence objects

jsiirola committed:
IndexedParam's initialize should treat sequences as mappings

jsiirola committed:
Add tests for initializing Set, Param from Series

jsiirola committed:
Change a Param test initializer that never made sense

Robbybp committed:
use calculate_variable_from_constraint when possible

blnicho committed:
Merge pull request #2128 from Robbybp/incidence-update

Updates to `incidence_analysis`

Robbybp committed:
Merge branch 'imp-fcn-hessian' into imp-fcn-performance

Robbybp committed:
add comments

Robbybp committed:
todo comments

bernalde commented:
Hi Zedong, this looks good. I agree with your responses. The only missing part is solving a couple of merging conflicts and pulling Pyomo/master into this branch to make sure that nothing has been affected by other recent PR's.
One final question, what do you mean when you want to do restructure the MindtPy_iteration_loop function ?

michaelbynum committed:
Merge pull request #2148 from michaelbynum/derivs

Common Sub-Expressions in Differentiation

jsiirola committed:
Merge branch 'main' into subsystem-ef

jsiirola committed:
Merge branch 'main' into expression-replacement

Robbybp committed:
remove commented code

jsiirola committed:
Remove reference to deprecated API

jsiirola committed:
NFC: fix typo

Commit:
Add option to force variables into nl file

Commit:
Update to least ampl_.py

jsiirola committed:
Merge pull request #2143 from jsiirola/expression-replacement

Rework expression replacement and LinearExpression object

Robbybp committed:
start tests for update hessian calculation

Robbybp committed:
dedicated function for setting external constraint multipliers

Commit:
Rename arg and add tests

jsiirola committed:
Add the ability to specify a custom Pint UnitRegistry

jsiirola committed:
Add a public attribute to access pint UnitRegistry

jsiirola committed:
Update (hidden) doctests to track PyomoUnitContainer API

Robbybp committed:
start test for full space hessians

Robbybp committed:
comment on potential slight refactor

Robbybp committed:
finish test for full space hessians

Robbybp committed:
nfc: remove whitespace

Robbybp committed:
test reduced hessian calculation

Robbybp committed:
test evaluate_hessian_equality_constraints

Robbybp committed:
move set_external_multipliers to hessian function so order of set_primals/set_duals does not matter

Robbybp committed:
test with order of set_primals/set_duals swapped

Robbybp committed:
Merge branch 'main' of https://github.com/Pyomo/pyomo into imp-fcn-performance

Robbybp committed:
remove old, now-unused hessian-of-lagrangian function

shermanjasonaf committed:
Add uncertain variable bounds detection to PyROS

Whenever PyROS replaces variable bounds with
explicit inequality constraints through the
`pyomo.contrib.pyros.util.turn_bounds_to_constraints` function,
the resulting constraint expressions do not contain
any mutable (and therefore, any uncertain) parameters
on which the bounds may depend, but the values of these parameters
at the point of bound/constraint replacement.
Hence, the `make_separation_problem` and
`make_separation_objectives` functions in
`pyomo.contrib.pyros.separation_problem_methods`
generate the corresponding separation problem objective
function expressions incorrectly for second-stage variables whose
bounds are uncertain.

To resolve: obtain the variable bound expressions through the
protected `_lb` and `_ub` attributes, and use these expressions
to generate explicit inequality constraints for all variables
whose bounds are contingent on uncertain parameters.
This constitutes a temporary fix, to be addressed further
with the use of a more secure getter/API in the future.

blnicho committed:
Merge pull request #2150 from jsiirola/df-initializer

Add `DataFrameInitializer`

emma58 committed:
PEP-8ing

emma58 committed:
Implementing the deprecation path for logical_to_linear not transforming BooleanVars that don't appear in expressions

Commit:
Add indexed variable support

Commit:
Merge branch 'main' of https://github.com/pyomo/pyomo into nl_force_var

emma58 committed:
Merge branch 'main' into l-to-l-targets

Commit:
Forgot to save a file

emma58 committed:
Addressing John's comments, all but where to move target_list

Robbybp committed:
change np.array to np.fromiter where applicable

Robbybp committed:
use name buffer where appropriate

Robbybp committed:
use unique_component_name and add test that fails without it

Robbybp committed:
skip test if ipopt not available

Robbybp committed:
do not pass list to extend

Robbybp committed:
use SolverFactory(ipopt)

Robbybp committed:
minor improvements to visitor

shermanjasonaf committed:
Make docstring Sphinx-compliant

jsiirola committed:
Add string representation for IndexedComponent_slice

jsiirola committed:
Add string representation for _ReferenceSet

jsiirola committed:
Update SetOf class hierarchy

- add support for InfiniteSetOf
- rename UnorderedSetOf to FiniteSetOf (to match other Set types)

jsiirola committed:
Add 'ReferenceTo' to pprint() of reference components

jsiirola committed:
Improve error when slicing component with UnknownSetDimen

This fixes #1928 (not by eliminating the error, but by explaining that
the behavior is not allowed and what the modeler can do to resolve it).

jsiirola committed:
Add test for error in #1928; suppress warning for set operators

jsiirola committed:
Fix #1800: be less aggressive with flattening indices

emma58 committed:
Deprecation path with longest class name in history

jsiirola committed:
Resolve error with InfiniteSetOf pprint() (fixes #2039)

jsiirola committed:
Support creating References to individual ComponentData objects

jsiirola committed:
NFC: minor doc/naming/import changes

Commit:
Fix redundant

jsiirola committed:
Resolve error converting nested single slices to a string

jsiirola committed:
NFC: updating documentation

Robbybp committed:
do not attach name buffer as instance attribute

jsiirola committed:
Improve Set test coverage

jsiirola committed:
Fix typo in __getstate__ (undefined symbol)

jsiirola committed:
NFC: whitespace and import cleanup

shermanjasonaf committed:
Add test for uncertain variable bounds/constraint replacement

shermanjasonaf committed:
Merge branch 'Pyomo:main' into main

emma58 committed:
Moving target_list function to core.util

emma58 committed:
NFC: Adding note for future Emma about Block-like targets

emma58 committed:
Fixing bug where when you legally associate a BooleanVar not used in any expressions after the transformation, you get yelled at

blnicho committed:
Merge pull request #2142 from Robbybp/subsystem-ef

Add external functions to block in `create_subsystem_block`

blnicho committed:
Merge pull request #2151 from jsiirola/pandas-set-param-init

Fixes for `Initializer(sequence)` and `pandas.Series`

Commit:
Fix _VarData oversight

ZedongPeng committed:
Merge branch 'main' into new-lp-nlp

emma58 committed:
Fixing bug in the error message about reassociating binary and Boolean vars

blnicho committed:
Merge pull request #2152 from eslickj/nl_force_var

Add option to force variables into nl file (for DAEs)

blnicho committed:
Merge pull request #2155 from Robbybp/imp-fcn-performance

`ExternalPyomoModel` performance improvements

jsiirola committed:
Add the exception to the NEOS connection failure warning message

jsiirola committed:
Add test for failed connection to NEOS

blnicho committed:
Merge pull request #2166 from jsiirola/neos-connection-error

Add the exception to the NEOS connection failure warning message

mrmundt committed:
Merge pull request #2158 from jsiirola/reference-bugfixes

Resolve Reference issues

emma58 committed:
Let targets be anything that claims to be a Block or is a Block (no requirements on whether ctype actually is a subclass of Block)

emma58 committed:
Correcting typo in logical to linear tests

emma58 committed:
Adding test for totally bizarre not-Component not-iterable target

jsiirola committed:
Update pint formatting for compatibility with pint 0.18

jsiirola committed:
Track change of xpress.range attribute in xpress 8.13.0

blnicho committed:
Merge pull request #2168 from jsiirola/pint-formatting-fix

Track changes in pint 0.18 and xpress 8.13.0

jsiirola committed:
Merge branch 'main' into l-to-l-targets

emma58 committed:
Changing deprecation version number to TBD

ZedongPeng committed:
increase code coverage

jsiirola committed:
Switch NumericRange to use +/-inf instead of None

jsiirola committed:
Tracking change to NumericRange use of +/-inf

jsiirola committed:
Remove unnecessary helper functions

jsiirola committed:
Fix tests that were asserting incorrect behavior

jsiirola committed:
Expand NumericRange tests

jsiirola committed:
NFC: document the __getattribute__ logic

jsiirola committed:
Restore original PyomoUnitContainer() behavior

Not specifying an argument now implicitly creates a new pint registry

jsiirola committed:
Minor ProductSet.bounds efficiency improvement

jsiirola committed:
Warn when constructing RangeSet with non-constant bounds

shermanjasonaf committed:
Improve efficiency of uncertain bound-to-constraint replacement

jsiirola committed:
Merge pull request #2153 from jsiirola/custom-pint-registry

Add support for custom pint registries

jsiirola committed:
Merge pull request #2170 from jsiirola/open-inf-numeric-range

Resolve bugs with open NumericRanges

jsiirola committed:
Add MaxExpression and MinExpression

jsiirola committed:
Make Var.free() a method so overriding unfix() also overrides free()

jsiirola committed:
Make Var.{lb,ub,bounds} settable, add Var.{lower,upper}

jsiirola committed:
Var bounds may not be potentially mutable

jsiirola committed:
Add tests for Var.{lower,upper}

jsiirola committed:
PEP8: avoid long lines

jsiirola committed:
NFC: documentation update

jsiirola committed:
Merge branch 'main' into var-bound-expr

jsiirola committed:
Standardize on Var.{lower,upper} as the abstract interface

jsiirola committed:
Update kernel to track changes in Var API

jsiirola committed:
Update GDP tests to track changes in Var API

ZedongPeng committed:
replace lambda with itemgetter

emma58 committed:
PEP-8ing line endings

emma58 committed:
PEP-8ing line endings

emma58 committed:
Checking if targets are instance of BlockData, this allows DisjunctData targets, which we need.

emma58 committed:
Changing logical to linear to not transform anything on Disjuncts--it ignores GDP components completely. (Though they can be targets.) Adding a comment soliloquy about this choice.

emma58 committed:
PEP-8ing GDPopt.py

emma58 committed:
PEP-8ing line endings in branch and bound

emma58 committed:
PEP-8ing line endings in config_options

emma58 committed:
line endings in cut_generation

emma58 committed:
Fixing line endings in the rest of gdpopt

emma58 committed:
Whoops, fixing a couple syntax errors

emma58 committed:
Fixing a couple ugly things

shermanjasonaf committed:
Add variable filter to `util.replace_uncertain_bounds_with_constraints`

shermanjasonaf committed:
Refine testing of uncertain bounds to constraint replacement

jsiirola committed:
NFC: fix deprecation message

ZedongPeng committed:
remove redundant code copy_var_list_values_from_solution_pool

emma58 committed:
Merge pull request #2147 from emma58/l-to-l-targets

Add support for solving Blocks and targets in core.logical_to_linear

emma58 committed:
Merge pull request #2173 from emma58/gdp-opt-pep8

NFC: PEP-8ing line endings in GDPopt

Robbybp committed:
allow CUID as argument to find_component

Robbybp committed:
test find_component with strings and cuids

jsiirola committed:
Resolve set performance degradation (introduced by #2170)

jsiirola committed:
Resolve performance degradation (var lb/ub getters)

jsiirola committed:
NFC: resolve comment typos & review requests

jsiirola committed:
Add testing of IndexedVar _varData lb/ub setters

Robbybp committed:
update find_component docstring

ZedongPeng committed:
correct typo

ZedongPeng committed:
improve copy_var_list_values and copy_var_list_values_from_solution_pool function

mrmundt committed:
Merge pull request #2179 from jsiirola/fix-set-perf-issue

Resolve set performance degradation (introduced by #2170)

mrmundt committed:
Merge pull request #2178 from Robbybp/find-component-cuid

Allow `find_component` to work with `ComponentUID`

mrmundt committed:
Resolving merge conflicts

mrmundt committed:
Resolving removed deprecation and RuntimeError

shermanjasonaf committed:
Add constraint replacement for variables in the objective

shermanjasonaf committed:
Check that variables/parameters added to new constraints are correct

jsiirola committed:
Merge pull request #2164 from ZedongPeng/new-lp-nlp

Add support for solution pool of MIP solvers in MindtPy

shermanjasonaf committed:
Generalize retrieval of variables from active model components

jsiirola committed:
Add debugging to identify failing mindtpy sub-test

jsiirola committed:
Send debugging output to stderr

jsiirola committed:
Skip test for Gurobi 9.5.0 on Windows

jsiirola committed:
Additional debugging information

jsiirola committed:
Fixing typo

jsiirola committed:
Fix model detection

jsiirola committed:
Skipping an additional test

jsiirola committed:
Skipping additional tests

jsiirola committed:
Skipping additional tests

jsiirola committed:
Skipping additional test

jsiirola committed:
Skipping additional test; refactor skip logic

blnicho committed:
Addressing comments from @jsiirola

jsiirola committed:
Removing debugging output

ZedongPeng committed:
add OA cuts after solving rNLP in FP standalone

jsiirola committed:
Merge pull request #2183 from jsiirola/mindtpy-debugging

Skip specific MindtPy tests due to failure with Gurobi 9.5.0 on Windows

mrmundt committed:
Merge branch 'main' of github.com:Pyomo/pyomo into deprecation-removal-for-6.0

mrmundt committed:
Merge branch 'deprecation-removal-for-6.0' of github.com:carldlaird/pyomo into deprecation-removal-for-6.0

mrmundt committed:
Merge branch 'main' of github.com:pyomo/pyomo into main

mrmundt committed:
Merge branch 'main' of github.com:pyomo/pyomo into new-lp-nlp

mrmundt committed:
Merge pull request #1992 from carldlaird/deprecation-removal-for-6.0

Deprecation removal for 6.0 - part I

jsiirola committed:
Remove unused import

jsiirola committed:
NFC: update docstrings

jsiirola committed:
Merge branch 'main' into var-bound-expr

jsiirola committed:
Fix typo in booktest skip declaration

jsiirola committed:
Merge pull request #2185 from ZedongPeng/new-lp-nlp

Fix MindtPy Feasibility Pump standalone bug

mrmundt committed:
Merge pull request #2172 from jsiirola/var-bound-expr

Public API to retrieve Var bound expressions

mrmundt committed:
Merge pull request #2186 from jsiirola/fix-book-test-skip

Fix typo in booktest skip declaration

jsiirola committed:
Merge branch 'main' into jason-main

jsiirola committed:
Uncertain bounds conversion: use public 'lower/upper' property

ZedongPeng committed:
ignore unused variables in get_integer_solution

ZedongPeng committed:
replace component_data_objects with variable constraint list

ZedongPeng committed:
Merge branch 'new-lp-nlp' of github.com:ZedongPeng/pyomo into new-lp-nlp

jsiirola committed:
Merge pull request #2159 from shermanjasonaf/main

Add uncertain variable bounds detection to PyROS

Robbybp committed:
specify dimension for sets with no entries (yet)

Robbybp committed:
do not use Set.__getitem__

Robbybp committed:
provide dimen for set with not entries in abstract model

Robbybp committed:
add within=Any to avoid deprecation warning

ZedongPeng committed:
improve add_solution

blnicho committed:
Merge pull request #2189 from Robbybp/fix-dae-examples

Fix errors and deprecation warnings in some DAE examples

ZedongPeng committed:
Merge branch 'main' into new-lp-nlp
